src/uct/cuda/cuda_copy/cuda_copy_ep.c

```diff
@@ -65,26 +65,60 @@ uct_cuda_copy_get_stream(uct_cuda_copy_ctx_rsc_t *ctx_rsc,
 }
 
 static UCS_F_ALWAYS_INLINE ucs_memory_type_t
-uct_cuda_copy_get_mem_type(uct_md_h md, void *address, size_t length)
+uct_cuda_copy_get_mem_type(uct_md_h md, void *address, size_t length,
+                           ucs_sys_device_t *sys_dev)
 {
     ucs_memory_info_t mem_info;
+    uct_md_mem_attr_t mem_attr;
     ucs_status_t status;
 
     status = ucs_memtype_cache_lookup(address, length, &mem_info);
     if (status == UCS_ERR_NO_ELEM) {
-        return UCS_MEMORY_TYPE_HOST;
+        goto out_host;
     }
 
     if ((status == UCS_ERR_UNSUPPORTED) ||
         (mem_info.type == UCS_MEMORY_TYPE_UNKNOWN)) {
-        status = uct_cuda_copy_md_detect_memory_type(md, address, length,
-                                                     &mem_info.type);
+        mem_attr.field_mask = UCT_MD_MEM_ATTR_FIELD_MEM_TYPE |
+                              UCT_MD_MEM_ATTR_FIELD_SYS_DEV;
+
+        status = uct_cuda_copy_md_mem_query(md, address, length, &mem_attr);
         if (status != UCS_OK) {
-            return UCS_MEMORY_TYPE_HOST;
+            goto out_host;
         }
+        mem_info.type    = mem_attr.mem_type;
+        mem_info.sys_dev = mem_attr.sys_dev;
     }
 
+    *sys_dev = mem_info.sys_dev;
     return mem_info.type;
+
+out_host:
+    *sys_dev = UCS_SYS_DEVICE_ID_UNKNOWN;
+    return UCS_MEMORY_TYPE_HOST;
+}
+
+static UCS_F_ALWAYS_INLINE void
+uct_cuda_copy_get_mem_types(uct_md_h md, void *src, void *dst, size_t length,
+                            ucs_memory_type_t *src_mem_type_p,
+                            ucs_memory_type_t *dst_mem_type_p,
+                            ucs_sys_device_t *sys_dev_p)
+{
+    ucs_sys_device_t src_sys_dev, dst_sys_dev;
+
+    *src_mem_type_p = uct_cuda_copy_get_mem_type(md, src, length, &src_sys_dev);
+    *dst_mem_type_p = uct_cuda_copy_get_mem_type(md, dst, length, &dst_sys_dev);
+    *sys_dev_p      = (src_sys_dev != UCS_SYS_DEVICE_ID_UNKNOWN) ?
+                      src_sys_dev : dst_sys_dev;
```

Comment:
It looks like we do not need to obtain everything for short operations. Maybe better to not unite everything in one function.